File: core/src/index/knowhere/knowhere/index/vector_index/ConfAdapter.cpp
// Copyright (C) 2019-2020 Zilliz. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions and limitations under the License.

#include "knowhere/index/vector_index/ConfAdapter.h"
#include <cmath>
#include <memory>
#include <string>
#include <vector>
#include "knowhere/common/Log.h"
#include "knowhere/index/vector_index/helpers/IndexParameter.h"

#ifdef MILVUS_GPU_VERSION
#include "faiss/gpu/utils/DeviceUtils.h"
#endif

namespace milvus {
namespace knowhere {

static const int64_t MIN_K = 0;
static const int64_t MAX_K = 16384;
static const int64_t MIN_NBITS = 1;
static const int64_t MAX_NBITS = 16;
static const int64_t DEFAULT_NBITS = 8;
static const int64_t MIN_NLIST = 1;
static const int64_t MAX_NLIST = 65536;
static const int64_t MIN_NPROBE = 1;
static const int64_t MAX_NPROBE = MAX_NLIST;
static const int64_t MIN_DIM = 1;
static const int64_t MAX_DIM = 32768;
static const int64_t HNSW_MIN_EFCONSTRUCTION = 8;
static const int64_t HNSW_MAX_EFCONSTRUCTION = 512;
static const int64_t HNSW_MIN_M = 4;
static const int64_t HNSW_MAX_M = 64;
static const int64_t HNSW_MAX_EF = 32768;
static const std::vector<std::string> FLT_METRICS{knowhere::Metric::L2, knowhere::Metric::IP};
static const std::vector<std::string> BIN_METRICS{Metric::HAMMING, Metric::JACCARD, Metric::TANIMOTO,
                                                  Metric::SUBSTRUCTURE, Metric::SUPERSTRUCTURE};
static const std::vector<std::string> BINIVF_METRICS{Metric::HAMMING, Metric::JACCARD, Metric::TANIMOTO};

#define CheckIntByRange(key, min, max)                                                                   \
    if (!oricfg.contains(key) || !oricfg[key].is_number_integer() || oricfg[key].get<int64_t>() > max || \
        oricfg[key].get<int64_t>() < min) {                                                              \
        return false;                                                                                    \
    }

#define CheckIntByRangeIfExist(key, min, max)                                       \
    if (oricfg.contains(key)) {                                                     \
        if (!oricfg[key].is_number_integer() || oricfg[key].get<int64_t>() > max || \
            oricfg[key].get<int64_t>() < min) {                                     \
            return false;                                                           \
        }                                                                           \
    }

#define CheckIntByValues(key, container)                                                                 \
    if (!oricfg.contains(key) || !oricfg[key].is_number_integer()) {                                     \
        return false;                                                                                    \
    } else {                                                                                             \
        auto finder = std::find(std::begin(container), std::end(container), oricfg[key].get<int64_t>()); \
        if (finder == std::end(container)) {                                                             \
            return false;                                                                                \
        }                                                                                                \
    }

#define CheckStrByValues(key, container)                                                                     \
    if (!oricfg.contains(key) || !oricfg[key].is_string()) {                                                 \
        return false;                                                                                        \
    } else {                                                                                                 \
        auto finder = std::find(std::begin(container), std::end(container), oricfg[key].get<std::string>()); \
        if (finder == std::end(container)) {                                                                 \
            return false;                                                                                    \
        }                                                                                                    \
    }

bool
ConfAdapter::CheckTrain(Config& oricfg, IndexMode& mode) {
    CheckIntByRange(meta::DIM, MIN_DIM, MAX_DIM);
    CheckStrByValues(Metric::TYPE, FLT_METRICS);
    return true;
}

bool
ConfAdapter::CheckSearch(Config& oricfg, const IndexType type, const IndexMode mode) {
    CheckIntByRange(meta::TOPK, MIN_K, MAX_K);
    return true;
}

int64_t
MatchNlist(int64_t size, int64_t nlist) {
    const int64_t MIN_POINTS_PER_CENTROID = 39;

    if (nlist * MIN_POINTS_PER_CENTROID > size) {
        // nlist is too large, adjust to a proper value
        nlist = std::max(1L, size / MIN_POINTS_PER_CENTROID);
        LOG_KNOWHERE_WARNING_ << "Row num " << size << " match nlist " << nlist;
    }
    return nlist;
}

int64_t
MatchNbits(int64_t size, int64_t nbits) {
    if (size < (1 << nbits)) {
        // nbits is too large, adjust to a proper value
        if (size >= (1 << 8)) {
            nbits = 8;
        } else if (size >= (1 << 4)) {
            nbits = 4;
        } else if (size >= (1 << 2)) {
            nbits = 2;
        } else {
            nbits = 1;
        }
        LOG_KNOWHERE_WARNING_ << "Row num " << size << " match nbits " << nbits;
    }
    return nbits;
}

bool
IVFConfAdapter::CheckTrain(Config& oricfg, IndexMode& mode) {
    CheckIntByRange(IndexParams::nlist, MIN_NLIST, MAX_NLIST);

    // auto tune params
    int64_t rows = oricfg[meta::ROWS].get<int64_t>();
    int64_t nlist = oricfg[IndexParams::nlist].get<int64_t>();
    oricfg[IndexParams::nlist] = MatchNlist(rows, nlist);

    return ConfAdapter::CheckTrain(oricfg, mode);
}

bool
IVFConfAdapter::CheckSearch(Config& oricfg, const IndexType type, const IndexMode mode) {
    if (mode == IndexMode::MODE_GPU) {
#ifdef MILVUS_GPU_VERSION
        CheckIntByRange(IndexParams::nprobe, MIN_NPROBE, faiss::gpu::getMaxKSelection());
#endif
    } else {
        CheckIntByRange(IndexParams::nprobe, MIN_NPROBE, MAX_NPROBE);
    }

    return ConfAdapter::CheckSearch(oricfg, type, mode);
}

bool
IVFSQConfAdapter::CheckTrain(Config& oricfg, IndexMode& mode) {
    oricfg[IndexParams::nbits] = DEFAULT_NBITS;
    return IVFConfAdapter::CheckTrain(oricfg, mode);
}

bool
IVFPQConfAdapter::CheckTrain(Config& oricfg, IndexMode& mode) {
    if (!IVFConfAdapter::CheckTrain(oricfg, mode)) {
        return false;
    }

    CheckIntByRangeIfExist(IndexParams::nbits, MIN_NBITS, MAX_NBITS);

    auto rows = oricfg[meta::ROWS].get<int64_t>();
    auto dimension = oricfg[meta::DIM].get<int64_t>();
    auto m = oricfg[IndexParams::m].get<int64_t>();
    auto nbits = oricfg.count(IndexParams::nbits) ? oricfg[IndexParams::nbits].get<int64_t>() : DEFAULT_NBITS;

    // auto tune params
    oricfg[IndexParams::nbits] = nbits = MatchNbits(rows, nbits);

#ifdef MILVUS_GPU_VERSION
    if (mode == IndexMode::MODE_GPU) {
        if (IsValidForGPU(dimension, m, nbits)) {
            return true;
        }
        // else try CPU Mode
        mode == IndexMode::MODE_CPU;
    }
#endif
    return IsValidForCPU(dimension, m);
}

bool
IVFPQConfAdapter::IsValidForGPU(int64_t dimension, int64_t m, int64_t nbits) {
    /*
     * Faiss 1.6
     * Only 1, 2, 3, 4, 6, 8, 10, 12, 16, 20, 24, 28, 32 dims per sub-quantizer are currently supported with
     * no precomputed codes. Precomputed codes supports any number of dimensions, but will involve memory overheads.
     */
    static std::vector<int64_t> support_dim_per_subquantizer{32, 28, 24, 20, 16, 12, 10, 8, 6, 4, 3, 2, 1};
    static std::vector<int64_t> support_subquantizer{96, 64, 56, 48, 40, 32, 28, 24, 20, 16, 12, 8, 4, 3, 2, 1};

    if (!IsValidForCPU(dimension, m)) {
        return false;
    }

    int64_t sub_dim = dimension / m;
    return (std::find(std::begin(support_subquantizer), std::end(support_subquantizer), m) !=
            support_subquantizer.end()) &&
           (std::find(std::begin(support_dim_per_subquantizer), std::end(support_dim_per_subquantizer), sub_dim) !=
                support_dim_per_subquantizer.end() &&
            (nbits == 8));
}

bool
IVFPQConfAdapter::IsValidForCPU(int64_t dimension, int64_t m) {
    return (dimension % m == 0);
}

bool
NSGConfAdapter::CheckTrain(Config& oricfg, IndexMode& mode) {
    static int64_t MIN_KNNG = 5;
    static int64_t MAX_KNNG = 300;
    static int64_t MIN_SEARCH_LENGTH = 10;
    static int64_t MAX_SEARCH_LENGTH = 300;
    static int64_t MIN_OUT_DEGREE = 5;
    static int64_t MAX_OUT_DEGREE = 300;
    static int64_t MIN_CANDIDATE_POOL_SIZE = 50;
    static int64_t MAX_CANDIDATE_POOL_SIZE = 1000;

    CheckIntByRange(IndexParams::knng, MIN_KNNG, MAX_KNNG);
    CheckIntByRange(IndexParams::search_length, MIN_SEARCH_LENGTH, MAX_SEARCH_LENGTH);
    CheckIntByRange(IndexParams::out_degree, MIN_OUT_DEGREE, MAX_OUT_DEGREE);
    CheckIntByRange(IndexParams::candidate, MIN_CANDIDATE_POOL_SIZE, MAX_CANDIDATE_POOL_SIZE);

    // auto tune params
    oricfg[IndexParams::nlist] = MatchNlist(oricfg[meta::ROWS].get<int64_t>(), 8192);

    int64_t nprobe = int(oricfg[IndexParams::nlist].get<int64_t>() * 0.1);
    oricfg[IndexParams::nprobe] = nprobe < 1 ? 1 : nprobe;

    return ConfAdapter::CheckTrain(oricfg, mode);
}

bool
NSGConfAdapter::CheckSearch(Config& oricfg, const IndexType type, const IndexMode mode) {
    static int64_t MIN_SEARCH_LENGTH = 1;
    static int64_t MAX_SEARCH_LENGTH = 300;

    CheckIntByRange(IndexParams::search_length, MIN_SEARCH_LENGTH, MAX_SEARCH_LENGTH);
    return ConfAdapter::CheckSearch(oricfg, type, mode);
}

bool
HNSWConfAdapter::CheckTrain(Config& oricfg, IndexMode& mode) {
    CheckIntByRange(IndexParams::efConstruction, HNSW_MIN_EFCONSTRUCTION, HNSW_MAX_EFCONSTRUCTION);
    CheckIntByRange(IndexParams::M, HNSW_MIN_M, HNSW_MAX_M);
    return ConfAdapter::CheckTrain(oricfg, mode);
}

bool
HNSWConfAdapter::CheckSearch(Config& oricfg, const IndexType type, const IndexMode mode) {
    CheckIntByRange(IndexParams::ef, oricfg[meta::TOPK], HNSW_MAX_EF);
    return ConfAdapter::CheckSearch(oricfg, type, mode);
}

bool
BinIDMAPConfAdapter::CheckTrain(Config& oricfg, IndexMode& mode) {
    CheckIntByRange(meta::DIM, MIN_DIM, MAX_DIM);
    CheckStrByValues(Metric::TYPE, BIN_METRICS);
    return true;
}

bool
BinIVFConfAdapter::CheckTrain(Config& oricfg, IndexMode& mode) {
    CheckIntByRange(meta::DIM, MIN_DIM, MAX_DIM);
    CheckStrByValues(Metric::TYPE, BINIVF_METRICS);
    CheckIntByRange(IndexParams::nlist, MIN_NLIST, MAX_NLIST);

    // auto tune params
    int64_t rows = oricfg[meta::ROWS].get<int64_t>();
    int64_t nlist = oricfg[IndexParams::nlist].get<int64_t>();
    oricfg[IndexParams::nlist] = MatchNlist(rows, nlist);
    return true;
}

bool
ANNOYConfAdapter::CheckTrain(Config& oricfg, IndexMode& mode) {
    static int64_t MIN_NTREES = 1;
    static int64_t MAX_NTREES = 1024;

    CheckIntByRange(IndexParams::n_trees, MIN_NTREES, MAX_NTREES);
    return ConfAdapter::CheckTrain(oricfg, mode);
}

bool
ANNOYConfAdapter::CheckSearch(Config& oricfg, const IndexType type, const IndexMode mode) {
    return ConfAdapter::CheckSearch(oricfg, type, mode);
}

}  // namespace knowhere
}  // namespace milvus

File: internal/util/indexparamcheck/conf_adapter.go
// Licensed to the LF AI & Data foundation under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package indexparamcheck

import (
	"fmt"
	"strconv"

	"github.com/milvus-io/milvus-proto/go-api/schemapb"
	"github.com/milvus-io/milvus/internal/util/funcutil"
)

type ConfAdapter interface {
	// CheckTrain returns true if the index can be built with the specific index parameters.
	CheckTrain(map[string]string) error
	CheckValidDataType(dType schemapb.DataType) bool
	CheckUnnecessaryParams(map[string]string) error
}

// BaseConfAdapter checks if a `FLAT` index can be built.
type BaseConfAdapter struct {
}

// CheckTrain check whether the params contains supported metrics types
func (adapter *BaseConfAdapter) CheckTrain(params map[string]string) error {
	if !CheckIntByRange(params, DIM, DefaultMinDim, DefaultMaxDim) {
		return fmt.Errorf("invild param with: %s", DIM)
	}

	return CheckStrByValues(params, Metric, METRICS)
}

// CheckValidDataType check whether the field data type is supported for the index type
func (adapter *BaseConfAdapter) CheckValidDataType(dType schemapb.DataType) bool {
	return true
}

func (adapter *BaseConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(FLATParams, params)
}

func newBaseConfAdapter() *BaseConfAdapter {
	return &BaseConfAdapter{}
}

// IVFConfAdapter checks if a IVF index can be built.
type IVFConfAdapter struct {
	BaseConfAdapter
}

// CheckTrain returns true if the index can be built with the specific index parameters.
func (adapter *IVFConfAdapter) CheckTrain(params map[string]string) error {
	if !CheckIntByRange(params, NLIST, MinNList, MaxNList) {
		return fmt.Errorf("invild param with: %s", NLIST)
	}

	// skip check number of rows

	return adapter.BaseConfAdapter.CheckTrain(params)
}

func (adapter *IVFConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(IVFFLATParams, params)
}

func newIVFConfAdapter() *IVFConfAdapter {
	return &IVFConfAdapter{}
}

// IVFPQConfAdapter checks if a IVF_PQ index can be built.
type IVFPQConfAdapter struct {
	IVFConfAdapter
}

// CheckTrain checks if ivf-pq index can be built with the specific index parameters.
func (adapter *IVFPQConfAdapter) CheckTrain(params map[string]string) error {
	if err := adapter.IVFConfAdapter.CheckTrain(params); err != nil {
		return err
	}

	return adapter.checkPQParams(params)
}

func (adapter *IVFPQConfAdapter) checkPQParams(params map[string]string) error {
	dimStr, dimensionExist := params[DIM]
	if !dimensionExist {
		return fmt.Errorf("lack necessary parameter: %s", DIM)
	}

	dimension, err := strconv.Atoi(dimStr)
	if err != nil { // invalid dimension
		return fmt.Errorf("invalid param with: %s", DIM)
	}

	// nbits can be set to default: 8
	nbitsStr, nbitsExist := params[NBITS]
	var nbits int
	if !nbitsExist {
		nbits = 8
	} else {
		nbits, err = strconv.Atoi(nbitsStr)
		if err != nil { // invalid nbits
			return fmt.Errorf("invalid param with: %s", NBITS)
		}
	}

	mStr, ok := params[IVFM]
	if !ok {
		return fmt.Errorf("lack necessary parameter: %s", IVFM)
	}
	m, err := strconv.Atoi(mStr)
	if err != nil || m == 0 { // invalid m
		return fmt.Errorf("invalid param with: %s", IVFM)
	}

	mode, ok := params[IndexMode]
	if !ok {
		mode = CPUMode
	}

	if mode == GPUMode && !adapter.checkGPUPQParams(dimension, m, nbits) {
		return fmt.Errorf("invalid param in %s mode", mode)
	}

	if !adapter.checkCPUPQParams(dimension, m) {
		return fmt.Errorf("invalid param in %s mode", mode)
	}

	return nil
}

func (adapter *IVFPQConfAdapter) checkGPUPQParams(dimension, m, nbits int) bool {
	/*
	 * Faiss 1.6
	 * Only 1, 2, 3, 4, 6, 8, 10, 12, 16, 20, 24, 28, 32 dims per sub-quantizer are currently supported with
	 * no precomputed codes. Precomputed codes supports any number of dimensions, but will involve memory overheads.
	 */

	subDim := dimension / m
	return funcutil.SliceContain(supportSubQuantizer, m) && funcutil.SliceContain(supportDimPerSubQuantizer, subDim) && nbits == 8
}

func (adapter *IVFPQConfAdapter) checkCPUPQParams(dimension, m int) bool {
	return (dimension % m) == 0
}

func (adapter *IVFPQConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(IVFPQParams, params)
}

func newIVFPQConfAdapter() *IVFPQConfAdapter {
	return &IVFPQConfAdapter{}
}

// IVFSQConfAdapter checks if a IVF_SQ index can be built.
type IVFSQConfAdapter struct {
	IVFConfAdapter
}

func (adapter *IVFSQConfAdapter) checkNBits(params map[string]string) bool {
	// cgo will set this key to DefaultNBits (8), which is the only value Milvus supports.
	_, exist := params[NBITS]
	if exist {
		// 8 is the only supported nbits.
		return CheckIntByRange(params, NBITS, DefaultNBits, DefaultNBits)
	}
	return true
}

// CheckTrain returns true if the index can be built with the specific index parameters.
func (adapter *IVFSQConfAdapter) CheckTrain(params map[string]string) error {
	if !adapter.checkNBits(params) {
		return fmt.Errorf("invild param with: %s", NBITS)
	}
	return adapter.IVFConfAdapter.CheckTrain(params)
}

func (adapter *IVFSQConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(IVFSQ8Params, params)
}

func newIVFSQConfAdapter() *IVFSQConfAdapter {
	return &IVFSQConfAdapter{}
}

type BinIDMAPConfAdapter struct {
	BaseConfAdapter
}

// CheckTrain checks if a binary flat index can be built with the specific parameters.
func (adapter *BinIDMAPConfAdapter) CheckTrain(params map[string]string) error {
	if !CheckIntByRange(params, DIM, DefaultMinDim, DefaultMaxDim) {
		return fmt.Errorf("invild param with: %s", DIM)
	}

	return CheckStrByValues(params, Metric, BinIDMapMetrics)
}

func (adapter *BinIDMAPConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(BINFLATParams, params)
}

func newBinIDMAPConfAdapter() *BinIDMAPConfAdapter {
	return &BinIDMAPConfAdapter{}
}

// BinIVFConfAdapter checks if a bin IFV index can be built.
type BinIVFConfAdapter struct {
	BaseConfAdapter
}

// CheckTrain checks if a binary ivf index can be built with specific parameters.
func (adapter *BinIVFConfAdapter) CheckTrain(params map[string]string) error {
	if !CheckIntByRange(params, DIM, DefaultMinDim, DefaultMaxDim) {
		return fmt.Errorf("invild param with: %s", DIM)
	}

	if !CheckIntByRange(params, NLIST, MinNList, MaxNList) {
		return fmt.Errorf("invild param with: %s", DIM)
	}

	// skip checking the number of rows

	return CheckStrByValues(params, Metric, BinIvfMetrics)
}

func (adapter *BinIVFConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(BINIVFFLATParams, params)
}

func newBinIVFConfAdapter() *BinIVFConfAdapter {
	return &BinIVFConfAdapter{}
}

type NSGConfAdapter struct {
	BaseConfAdapter
}

// CheckTrain checks if a nsg index can be built with specific parameters.
func (adapter *NSGConfAdapter) CheckTrain(params map[string]string) error {
	if err := CheckStrByValues(params, Metric, METRICS); err != nil {
		return err
	}

	if !CheckIntByRange(params, KNNG, MinKNNG, MaxKNNG) {
		return fmt.Errorf("invild param with: %s", KNNG)
	}

	if !CheckIntByRange(params, SearchLength, MinSearchLength, MaxSearchLength) {
		return fmt.Errorf("invild param with: %s", SearchLength)
	}

	if !CheckIntByRange(params, OutDegree, MinOutDegree, MaxOutDegree) {
		return fmt.Errorf("invild param with: %s", OutDegree)
	}

	if !CheckIntByRange(params, CANDIDATE, MinCandidatePoolSize, MaxCandidatePoolSize) {
		return fmt.Errorf("invild param with: %s", CANDIDATE)
	}

	// skip checking the number of rows

	return nil
}

func (adapter *NSGConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(NSGParams, params)
}

func newNSGConfAdapter() *NSGConfAdapter {
	return &NSGConfAdapter{}
}

// HNSWConfAdapter checks if a hnsw index can be built.
type HNSWConfAdapter struct {
	BaseConfAdapter
}

// CheckTrain checks if a hnsw index can be built with specific parameters.
func (adapter *HNSWConfAdapter) CheckTrain(params map[string]string) error {
	if !CheckIntByRange(params, EFConstruction, HNSWMinEfConstruction, HNSWMaxEfConstruction) {
		return fmt.Errorf("invild param with: %s", EFConstruction)
	}

	if !CheckIntByRange(params, HNSWM, HNSWMinM, HNSWMaxM) {
		return fmt.Errorf("invild param with: %s", HNSWM)
	}

	return adapter.BaseConfAdapter.CheckTrain(params)
}

func (adapter *HNSWConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(HNSWParams, params)
}

func newHNSWConfAdapter() *HNSWConfAdapter {
	return &HNSWConfAdapter{}
}

// ANNOYConfAdapter checks if an ANNOY index can be built.
type ANNOYConfAdapter struct {
	BaseConfAdapter
}

// CheckTrain checks if an annoy index can be built with specific parameters.
func (adapter *ANNOYConfAdapter) CheckTrain(params map[string]string) error {
	if !CheckIntByRange(params, NTREES, MinNTrees, MaxNTrees) {
		return fmt.Errorf("invild param with: %s", NTREES)
	}

	return adapter.BaseConfAdapter.CheckTrain(params)
}

func (adapter *ANNOYConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(ANNOYParams, params)
}

func newANNOYConfAdapter() *ANNOYConfAdapter {
	return &ANNOYConfAdapter{}
}

// RHNSWFlatConfAdapter checks if a rhnsw flat index can be built.
type RHNSWFlatConfAdapter struct {
	HNSWConfAdapter
}

// CheckTrain checks if a rhnsw flat index can be built with specific parameters.
func (adapter *RHNSWFlatConfAdapter) CheckTrain(params map[string]string) error {
	return adapter.HNSWConfAdapter.CheckTrain(params)
}

func (adapter *RHNSWFlatConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(HNSWParams, params)
}

func newRHNSWFlatConfAdapter() *RHNSWFlatConfAdapter {
	return &RHNSWFlatConfAdapter{}
}

// RHNSWPQConfAdapter checks if a rhnsw pq index can be built.
type RHNSWPQConfAdapter struct {
	IVFPQConfAdapter
	HNSWConfAdapter
}

// CheckTrain checks if a rhnsw pq index can be built with specific parameters.
func (adapter *RHNSWPQConfAdapter) CheckTrain(params map[string]string) error {
	if err := adapter.HNSWConfAdapter.CheckTrain(params); err != nil {
		return err
	}

	dimension, _ := strconv.Atoi(params[DIM])
	pqmStr, ok := params[PQM]
	if !ok {
		return fmt.Errorf("lack necessary param: %s", PQM)
	}
	pqm, err := strconv.Atoi(pqmStr)
	if err != nil || pqm == 0 {
		return fmt.Errorf("invild param with: %s", PQM)
	}

	if !adapter.IVFPQConfAdapter.checkCPUPQParams(dimension, pqm) {
		return fmt.Errorf("invild param with: %s", PQM)
	}

	return nil
}

func (adapter *RHNSWPQConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(RHNSWPQParams, params)
}

func newRHNSWPQConfAdapter() *RHNSWPQConfAdapter {
	return &RHNSWPQConfAdapter{}
}

// RHNSWSQConfAdapter checks if a rhnsw sq index can be built.
type RHNSWSQConfAdapter struct {
	BaseConfAdapter
	HNSWConfAdapter
}

// CheckTrain checks if a rhnsw sq index can be built with specific parameters.
func (adapter *RHNSWSQConfAdapter) CheckTrain(params map[string]string) error {
	return adapter.HNSWConfAdapter.CheckTrain(params)
}

func (adapter *RHNSWSQConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(HNSWParams, params)
}

func newRHNSWSQConfAdapter() *RHNSWSQConfAdapter {
	return &RHNSWSQConfAdapter{}
}

// NGTPANNGConfAdapter checks if a NGT_PANNG index can be built.
type NGTPANNGConfAdapter struct {
	BaseConfAdapter
}

func (adapter *NGTPANNGConfAdapter) CheckTrain(params map[string]string) error {
	if !CheckIntByRange(params, EdgeSize, NgtMinEdgeSize, NgtMaxEdgeSize) {
		return fmt.Errorf("invild param with: %s", EdgeSize)
	}

	if !CheckIntByRange(params, ForcedlyPrunedEdgeSize, NgtMinEdgeSize, NgtMaxEdgeSize) {
		return fmt.Errorf("invild param with: %s", ForcedlyPrunedEdgeSize)
	}

	if !CheckIntByRange(params, SelectivelyPrunedEdgeSize, NgtMinEdgeSize, NgtMaxEdgeSize) {
		return fmt.Errorf("invild param with: %s", SelectivelyPrunedEdgeSize)
	}

	selectivelyPrunedEdgeSize, _ := strconv.Atoi(params[SelectivelyPrunedEdgeSize])
	forcedlyPrunedEdgeSize, _ := strconv.Atoi(params[ForcedlyPrunedEdgeSize])
	if selectivelyPrunedEdgeSize >= forcedlyPrunedEdgeSize {
		return fmt.Errorf("invild param with: %s and %s", ForcedlyPrunedEdgeSize, ForcedlyPrunedEdgeSize)
	}

	return adapter.BaseConfAdapter.CheckTrain(params)
}

func (adapter *NGTPANNGConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(NGTPANNGParams, params)
}

func newNGTPANNGConfAdapter() *NGTPANNGConfAdapter {
	return &NGTPANNGConfAdapter{}
}

// NGTONNGConfAdapter checks if a NGT_ONNG index can be built.
type NGTONNGConfAdapter struct {
	BaseConfAdapter
}

func (adapter *NGTONNGConfAdapter) CheckTrain(params map[string]string) error {
	if !CheckIntByRange(params, EdgeSize, NgtMinEdgeSize, NgtMaxEdgeSize) {
		return fmt.Errorf("invild param with: %s", EdgeSize)
	}

	if !CheckIntByRange(params, OutgoingEdgeSize, NgtMinEdgeSize, NgtMaxEdgeSize) {
		return fmt.Errorf("invild param with: %s", OutgoingEdgeSize)
	}

	if !CheckIntByRange(params, IncomingEdgeSize, NgtMinEdgeSize, NgtMaxEdgeSize) {
		return fmt.Errorf("invild param with: %s", IncomingEdgeSize)
	}

	return adapter.BaseConfAdapter.CheckTrain(params)
}

func (adapter *NGTONNGConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(NGTONNGParams, params)
}

func newNGTONNGConfAdapter() *NGTONNGConfAdapter {
	return &NGTONNGConfAdapter{}
}

type DISKANNConfAdapter struct {
	BaseConfAdapter
}

func (adapter *DISKANNConfAdapter) CheckTrain(params map[string]string) error {
	if !CheckIntByRange(params, DIM, DiskAnnMinDim, DiskAnnMaxDim) {
		return fmt.Errorf("invild param with: %s", DIM)
	}
	return nil
}

func (adapter *DISKANNConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(DISKANNParams, params)
}

// CheckValidDataType check whether the field data type is supported for the index type
func (adapter *DISKANNConfAdapter) CheckValidDataType(dType schemapb.DataType) bool {
	vecDataTypes := []schemapb.DataType{
		schemapb.DataType_FloatVector,
	}
	return funcutil.SliceContain(vecDataTypes, dType)
}

func newDISKANNConfAdapter() *DISKANNConfAdapter {
	return &DISKANNConfAdapter{}
}

type ScalarConfAdapter struct {
	BaseConfAdapter
}

func (adapter *ScalarConfAdapter) CheckTrain(params map[string]string) error {
	return nil
}

func (adapter *ScalarConfAdapter) CheckUnnecessaryParams(params map[string]string) error {
	return checkUnnecessaryParams(ScalarIndexParams, params)
}

func newScalarConfAdapter() *ScalarConfAdapter {
	return &ScalarConfAdapter{}
}
